codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java

```diff
@@ -83,9 +85,14 @@
         void onWeightChanged(Http2Stream stream, short oldWeight);
 
         /**
-         * Called when a GO_AWAY frame has either been sent or received for the connection.
+         * Called when a GO_AWAY frame was sent for the connection.
+         */
+        void goAwaySent(long errorCode, ByteBuf debugData);
+
+        /**
+         * Called when a GO_AWAY frame was received for the connection.
          */
-        void goingAway();
+        void goAwayReceived(long errorCode, ByteBuf debugData);
```

Comment:
@Scottmitch sorry, just getting back to this discussion...

So we have 2 interfaces: `Http2Connection.Listener` and `Http2FrameListener`. The latter being a low-level interface where the application can get inbound DATA, HEADERS, etc. The former providing more abstract events about the state of the connection (usually triggered by the low-level events).

`GOAWAY` is a low-level event that directly translates to a change in connection state, which is why we're having so much trouble here :).  It's clear that it has to be part of `Http2FrameListener`, but the question is whether it should also be a part of `Http2Connection.Listener` and if so, how?

The `Http2Connection` interface already had getter/setter for GOAWAY sent/received since it is a general piece of connection state used in various places. I don't think we should change this.  `Http2Connection.Listener` also had events for when these values in the connection were set.  I guess we can ask what value are these events adding? I would say that they are useful and not entirely redundant with `Http2FrameListener.onGoAwayRead` for the following reasons:

1)  Having these events in the `Http2Connection.Listener` allows various parts of the application to react to GOAWAY.  In contrast, if the application relied on the sole `Http2FrameListener` instance for the GOAWAY event, they would have to manually distribute the GOAWAY event to every place that needs it.

2) Having these events in the `Http2Connection.Listener` also provides a one-stop-shop listener interface for getting notified of both inbound and outbound GOAWAY. Applications could then react to GOAWAY (inbound or outbound) in a simple, unified way. Capturing outbound GOAWAY in a generic way would currently have to be done by intercepting the call to the `Http2FrameWriter`, which seems hacky.

So I guess I still feel that the redundancy isn't all that redundant. We could potentially remove the "redundancy" by just not calling that `onGoAwayRead` method from the decoder (and document that of course).

WDYT?